plutus-tx-plugin/test/List/Spec.hs

```diff
@@ -0,0 +1,447 @@
+        $ mapMaybeProgram
+        `unsafeApplyCode` liftCodeDef num
+        `unsafeApplyCode` liftCodeDef list
+    )
+    (===)
+    (semanticsToList expected)
+  cekResultMatchesHaskellValue
+    ( compiledCodeToTerm
+        $ dataMapMaybeProgram
+        `unsafeApplyCode` liftCodeDef num
+        `unsafeApplyCode` liftCodeDef dataList
+    )
+    (===)
+    (semanticsToList expected)
+
+anyProgram :: CompiledCode (Integer -> [Integer] -> Bool)
+anyProgram = $$(compile [|| \n -> List.any (\x -> x PlutusTx.> n) ||])
+
+dataAnyProgram  :: CompiledCode (Integer -> Data.List Integer -> Bool)
+dataAnyProgram = $$(compile [|| \n -> Data.List.any (\x -> x PlutusTx.> n) ||])
+
+anySpec :: Property
+anySpec = property $ do
+  listS <- forAll genListS
+  num <- forAll $ Gen.integral rangeElem
+  let list = semanticsToList listS
+      dataList = semanticsToDataList listS
+      expected = anyS (> num) listS
+  cekResultMatchesHaskellValue
+    ( compiledCodeToTerm
+        $ anyProgram
+        `unsafeApplyCode` liftCodeDef num
+        `unsafeApplyCode` liftCodeDef list
+    )
+    (===)
+    expected
+  cekResultMatchesHaskellValue
+    ( compiledCodeToTerm
+        $ dataAnyProgram
+        `unsafeApplyCode` liftCodeDef num
+        `unsafeApplyCode` liftCodeDef dataList
+    )
+    (===)
+    expected
+
+foldMapProgram :: CompiledCode (Integer -> [Integer] -> Maybe [Integer])
+foldMapProgram =
+  $$(compile
+  [||
+    \n -> PlutusTx.foldMap (\x -> if x PlutusTx.> n then Just [x] else Nothing)
+  ||])
+
+dataFoldMapProgram :: CompiledCode (Integer -> Data.List Integer -> Maybe [Integer])
+dataFoldMapProgram =
+  $$(compile
+  [||
+    \n ->  Data.List.foldMap (\x -> if x PlutusTx.> n then Just [x] else Nothing)
+  ||])
+
+foldMapSpec :: Property
+foldMapSpec = property $ do
+  listS <- forAll genListS
+  num <- forAll $ Gen.integral rangeElem
+  let list = semanticsToList listS
+      dataList = semanticsToDataList listS
+      expected = foldMapS (\x -> if x > num then Just [x] else Nothing) listS
+  cekResultMatchesHaskellValue
+    ( compiledCodeToTerm
+        $ foldMapProgram
+        `unsafeApplyCode` liftCodeDef num
+        `unsafeApplyCode` liftCodeDef list
+    )
+    (===)
+    expected
+  cekResultMatchesHaskellValue
+    ( compiledCodeToTerm
+        $ dataFoldMapProgram
+        `unsafeApplyCode` liftCodeDef num
+        `unsafeApplyCode` liftCodeDef dataList
+    )
+    (===)
+    expected
+
+mapProgram :: CompiledCode (Integer -> [Integer] -> [Integer])
+mapProgram = $$(compile [|| \n -> List.map (\x -> x PlutusTx.+ n) ||])
+
+dataMapProgram :: CompiledCode (Integer -> Data.List Integer -> Data.List Integer)
+dataMapProgram = $$(compile [|| \n -> Data.List.map (\x -> x PlutusTx.+ n) ||])
+
+mapSpec :: Property
+mapSpec = property $ do
+  listS <- forAll genListS
+  num <- forAll $ Gen.integral rangeElem
+  let list = semanticsToList listS
+      dataList = semanticsToDataList listS
+      expected = mapS (+ num) listS
```

Comment:
With `QuickCheck` you can generate actual functions. `hedgehog` is probably also capable of that.